### DISCUSSION TITLE
simulate trajectories based on individual-level estimates

### DISCUSSION THREAD

**Discussion by xiaoyang-mindstrong** [nlmixrdevelopment/nlmixr#608] [Ideas]
Hi all, 

(Disclaimer: I am new to the nlmixr package and have some experience in using nlme and lmer to build multilevel models.)

**Goal**: 
I am trying to simulate trajectories based on individual-level estimates, and I haven't found a ready-to-use function from the package yet. So in order to use the "nlmixrSim" function which takes the sample-level estimates, I was manipulating the model fit object in a way that pretends the individual estimates as the sample-level estimates. 

**Barrier**: 
The model fit object does not allow me to change the sample-level estimates. The code is as follows: 
model_fit_obj$coefficients$fixed <- individual_coeff
the error I got as follows:
Error:
! Assigned data `<named list>` must be compatible with existing data.
x Existing data has 1843 rows.
x Assigned data has 2 rows.
ℹ Only vectors of size 1 are recycled.
Backtrace:
  1. base::`$<-`(`*tmp*`, coefficients, value = `<named list>`)
 12. tibble `<fn>`(`<vctrs___>`)

**Question**: 
Why does the error tell me the existing data (which is the fixed effects of the model object) have 1843 rows (which is the number of all observations for all subjects), while the fixed effects should have only 6 coefficients? Is there a way I can fix this error? 

Thank you very much!

**Comment by mattfidler** [comment]
Try `augPred()` for your problem, you can also access simulation information for `RxODE` with `$simInfo`

**Reply by xiaoyang-mindstrong** [reply]
Hey Matt, 

Thanks for the quick reply and suggestion. I tried the augPred() function, and I wonder if the augPred function can take a customized dosage schedule like nlmixrSim() in some way. 

I tried to add one event table as a parameter, 
simulated <- augPred(fit , events= therapy_dose)

but got the following error
Error in nlmixrAugPred(object = object, primary = primary, minimum = minimum, : 
object '_nlmixr_augPredTrans' not found

Thank you!

**Reply by mattfidler** [reply]
This gives an example of simulating a new scenario:

https://nlmixr2.github.io/nlmixr2/articles/xgxr-nlmixr-ggpmx.html#simulation-of-a-new-scenario-with-rxode2

This should be helpful in your scenario

**Reply by xiaoyang-mindstrong** [reply]
Hey Matt, 
Thanks for the feedback. I looked at this link you suggested and it looks like it can do a hypothetical dosage schedule at the sample-level estimates with the number of subjects (e.g., 100. Correct me if I am wrong: this is not specific individuals in the sample, just a number of subjects to be simulated. If I am wrong about this, please let me know how I should specify the specific subject IDs that I want to simulate in the function). Sorry if this was not clear in my previous questions, and what I need is a function that can simulate a hypothetical dosage schedule based on individual-level estimates so that we can compute a personalized intervention plan for a specific person based on their historic data.

**Reply by mattfidler** [reply]
Hi @xiaoyang-mindstrong 

The original tutorial had an example of personalized dosing:

https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4728294/

In `nlmixr` and `nlmixr2` you can get the underlying `RxODE` and `rxode2` code from `fit$simInfo` and apply the individualized dosing as you wish.

**Reply by mattfidler** [reply]
The code they had there is:

```r
effect.limits = c(0, 0.4, 0.6, 9) # Decision rule limits

dose.multipliers = c(0.5, 1, 2) # Decision rule effects

ndays <‐ 25; unit.dose <‐ 10000; start.dose <‐ 1; sampling.frequency <‐ 1 # Sample every day

# Simulate each day. At the end of each day, test the effect level, and adjust the dose level according to the decision rule

for (i in seq(1, ndays, by = sampling.frequency)) {

if (i==1) {# Initialize on first day

inits <‐ c(0, 0, 0, 1)

last.multiplier <‐ start.dose

this.multiplier <‐ 1

} else {# Use end of previous day as initial conditions

for next day, compare trough effect with

#decision rule limits and determine dose multiplier accordingly

inits <‐ x[dim(x)[1], vars]

wh <‐ cut(inits[“eff”], effect.limits)

this.multiplier <‐ dose.multipliers[wh]

}

this.multiplier <‐ this.multiplier*last.multiplier # Adjust dose

last.multiplier <‐ this.multiplier# Store new dose

#specify dosing and sampling

ev <‐ eventTable()

ev$add.dosing(dose = this.multiplier*unit.dose, dosing.interval = 24, nbr.doses = sampling.frequency)

ev$add.sampling(0:(24*sampling.frequency))

x <‐ mod1$run(params, ev, inits) # Run simulation

# Compile outputs

time.total <‐ ev$get.EventTable()[,”time”]+(i‐1)*(24)

doses <‐ rep(last.multiplier, length(time))

x <‐ cbind(x, time.total, doses);

res <‐ rbind(res, x)

}
```

**Reply by xiaoyang-mindstrong** [reply]
hey Matt, sorry for the late reply. I have tried your suggestion and it worked. Thank you so much for your patience and guidance!
